cache/remotecache/export.go

```diff
@@ -43,18 +42,77 @@ const (
 	ExporterResponseManifestDesc = "cache.manifest"
 )
 
-type contentCacheExporter struct {
-	solver.CacheExporterTarget
-	chains   *v1.CacheChains
-	ingester content.Ingester
-	oci      bool
-	ref      string
-	comp     compression.Config
+func NewExporter(ingester content.Ingester, ref string, oci bool, imageManifest bool, compressionConfig compression.Config) Exporter {
+	cc := v1.NewCacheChains()
+	return &contentCacheExporter{CacheExporterTarget: cc, chains: cc, ingester: ingester, oci: oci, imageManifest: imageManifest, ref: ref, comp: compressionConfig}
 }
 
-func NewExporter(ingester content.Ingester, ref string, oci bool, compressionConfig compression.Config) Exporter {
-	cc := v1.NewCacheChains()
-	return &contentCacheExporter{CacheExporterTarget: cc, chains: cc, ingester: ingester, oci: oci, ref: ref, comp: compressionConfig}
+type ExportableCache struct {
+	exportedManifest ocispecs.Manifest
+	exportedIndex    ocispecs.Index
+	cacheType        int
+}
+
+func (ec *ExportableCache) GetMediaType() string {
+	if ec.cacheType == 0 {
+		return ec.exportedIndex.MediaType
+	}
+	return ec.exportedManifest.MediaType
+}
+
+func (ec *ExportableCache) SetSchemaVersion(version int) {
+	if ec.cacheType == 0 {
+		ec.exportedIndex.SchemaVersion = version
+	} else {
+		ec.exportedManifest.SchemaVersion = version
+	}
+}
+func (ec *ExportableCache) SetMediaType(mediaType string) {
+	if ec.cacheType == 0 {
+		ec.exportedIndex.MediaType = mediaType
+	} else {
+		ec.exportedManifest.MediaType = mediaType
+	}
+}
+
+func (ec *ExportableCache) AddCacheBlob(blob ocispecs.Descriptor) {
+	if ec.cacheType == 0 {
+		ec.exportedIndex.Manifests = append(ec.exportedIndex.Manifests, blob)
+	} else {
+		ec.exportedManifest.Layers = append(ec.exportedManifest.Layers, blob)
+	}
+}
+
+func (ec *ExportableCache) FinalizeCache(ctx context.Context, ce *contentCacheExporter) {
+	// Nothing needed here for Manifest-type cache manifests
```

Comment:
Are you sure `ConvertAllLayerMediaTypes` is not needed? Is it guaranteed to be OCI already?

Reply:
From my tests I think it "accidentally" was OCI compatible, but I think you're right in that there's no guarantee that whatever the solver gives is going to be a consistent OCI `Layer` type. We should call this regardless of cache type to ensure the entire cache manifest is internally consistent.

Reply:
Comment in here was not removed

Reply:
Literally had my IDE open to remove it as I got this message :)